### ISSUE TITLE
`hash` is present in the `URL` passed to `load`

### ISSUE THREAD

**Issue by tcc-sejohnson** [sveltejs/kit#4788]
### Describe the bug

In working on #4773, @benmccann stated a couple of behaviors:

(For the duration of this post, `fragment === hash`.)

1. `prefetch` should not rerun for identical routes with different hashes (eg. `/prefetch` and `/prefetch#some-hash`).
2. People should _not_ be using the `URL` `hash` property in `load`.

I would tend to agree that either both or neither of these statements should be true; if 1 then 2 OR if not 1 then not 2. The promise of `prefetch` is that `load` will be run -- if the `hash` is meant to be part of the `URL` supplied to `load`, then `load` should rerun when the hash changes. If the `hash` should _not_ be part of the `URL` supplied to `load`, `load` should not rerun when the `hash` changes.

If 1 is correct, [this branch](https://github.com/tcc-sejohnson/kit/tree/sejohnson-remove-hash-from-load-url) will take care of 2. (It was very easy to change and test.)

However, I'm having trouble with the justification of 1 and 2. Looking at the [URL spec](https://url.spec.whatwg.org/#concept-url-fragment), I can't find anywhere that says "the content of the document should not change" based on the fragment. [The HTML spec](https://html.spec.whatwg.org/multipage/browsing-the-web.html#scroll-to-fragid) also doesn't seem to require that the content of the page remain static regardless of the fragment. (Please note, I'm not an expert in the nitty gritty of the URL or HTML specs, so please correct me if I'm wrong and I'll edit this part to include a correction.) The URL spec defines the fragment as "[a] [string](https://infra.spec.whatwg.org/#ascii-string) that can be used for further processing on the resource the [URL](https://url.spec.whatwg.org/#concept-url)’s other components identify." The phrase "can be used for further processing on the resource" in the context of a web document would say to me "...can be used for further processing by JavaScript...", which would suggest we would want to enable users to have access to it in `load`.

The very existence of things like `ReactRouter`'s [`HashRouter`](https://v5.reactrouter.com/web/api/HashRouter) seem to go against the idea that the document's content should not change based on the hash.

I'm opening this issue as a place to discuss further. If we commit to both 1 and 2, this will be fixed by opening a pull request against the previously-linked branch. If we commit to not 1 and not 2, this will be fixed by leaving 2 alone (exposing the hash to `load` is the current behavior), and a small modification to #4477 to fix 1. I'm most concerned with SvelteKit continuing to align with expected spec behavior (and the expectations of its users)!

### Reproduction

N/A.

### Logs

```shell
N/A.
```

### System Info

```shell
N/A.
```

### Severity

annoyance

**Comment by Conduitry** [comment]
The `#` portion of URLs is not sent to the server by the browser, and so cannot be used as part of determining what's rendered during SSR for that page. When `load` is run on the browser, `hash` is available, because we are using the native `location` object to construct the URL. I would probably tend to say that we should _not_ artificially strip certain data out of the `location` when calling the `load` function, but if that necessarily means that people will be confused during SSR, that might force our hand.

SvelteKit also needs to decide whether it sees itself as SSR-first (which is likely something it already sees itself as), which would also answer the question of whether hash-only navigations should be considered. The HTML5 history API has existed for a long time, and - even for SPAs - SvelteKit would likely encourage you to use URLs with different `pathname`s to refer to different pages. If you want to have a single `pathname` but different `hash` values to refer to different pages, you can use a Svelte-only solution with no SvelteKit in sight.

**Comment by tcc-sejohnson** [comment]
> I would probably tend to say that we should not artificially strip certain data out of the location when calling the load function, but if that necessarily means that people will be confused during SSR, that might force our hand.

I would agree that we shouldn't artificially strip data as a general principle. I could see this _potentially_ being confusing during SSR, but I don't think it _should_ be, at least, not to any experienced web developer. Like you said, the `fragment`/`hash` portion of the URL is not sent to the server -- this is well-known behavior. Obviously, any references to `hash` in `load` would need to be wrapped in `if (browser)` (or handle the empty hash some other way).

> If you want to have a single pathname but different hash values to refer to different pages, you can use a Svelte-only solution with no SvelteKit in sight.

I don't think this is the only use case, though. It doubtless is the _majority_ use case, but say I have a regularly-updated page with lots of headings:

```html
<h1 id="adventurer-blog">I'm an adventurer</h1>
<h2 id="about-me">About me</h2>
<h2 id="my-strong-legs">My legs are swole</h2>
```

Then, later, as we all know is inevitable, you take an arrow to the knee and need to retract `my-strong-legs` with `my-terrible-injury`. The SvelteKit-y way to do that, IMO, would be to use `load`:

```ts
<script context="module" lang="ts">
  const hashRedirects = {
    "my-strong-legs": "my-terrible-injury",
  }

  export const load = ({ url }) => {
    hashRedirect = hashRedirects[url.hash];
    if (hashRedirect) {
      url.hash = hashRedirect;
      return {
        status: 301,
        redirect: url.href,
    }

    return {}
  }
</script>
```

Granted, there are other ways to do this, but there are other ways to do _most_ things you can do in `load`. 

I could also see removing the hash on the client negatively impacting our friends who run in SPA mode or use `adapter-static`, potentially?

**Comment by Mlocik97** [comment]
actually it's good thing it's like this... I even use that hash often... but if you don't need it, it's easy to extract (resp remove it) from URL, But otherway, it would be problem...

**Comment by tcc-sejohnson** [comment]
The more I think about this, the more I side with Ben's original comment. While I think the URL and HTTP specs leave a lot of room for interpretation as to what the hash can/should be used for (the URL spec practically _begs_ for JavaScript to be able to use it "for further processing"), I feel more and more strongly it should be kept out of `load`, even artificially, though it _should_ be correctly supplied in the `page` store on navigation after `prefetch` has run.

I suppose my final vote would be to merge #4773 in order to fix the `page` store behavior and prevent `prefetch` from running multiple times for identical routes with different hashes, then opening a pull request against  https://github.com/tcc-sejohnson/kit/tree/sejohnson-remove-hash-from-load-url to remove the `hash` from the URL `load` receives. I'll wait on more feedback from the maintainers before opening that second PR.

**Comment by Rich-Harris** [comment]
There's another consideration here: the behaviour when the user clicks a link with a hash fragment.

Right now, SvelteKit's router bails out when you click `<a href="#some-fragment">` — it updates the scroll position history and updates the `page` store, but that's all. It doesn't re-run `load` functions.

So in the case where someone _is_ using `url.hash` in `load` for some reason...

```js
return {
  props: {
    hash: url.hash
  }
}
```

...the page will be incorrect. There are two options:

1. make `url.hash` inaccessible in `load`
2. re-run `load` when clicking hash links

Anything else (including the status quo) is broken IMHO. 

Re-running `load` feels too costly, which makes me feel like 1 is a better option, particularly given the other arguments (`prefetch` behaviour and server/client consistency). We _could_ rerun `load` if and only if `url.hash` was accessed, but we can't detect someone accessing the hash via `url.href` (or `url.toString()`). (Maybe we treat accessing `url.href` the same as accessing `url.hash` for this purpose?)

At the same time, people _will_ get annoyed if `url.hash` is artifically inaccessible in `load`. All of which is a longwinded way of saying that I'm not really sure what we should do here.

**Comment by Mlocik97** [comment]
for me it doesn't make sense to re-run load when hash change / or is accessed, as it's part of URL that shoudln't have effect on content of page, it should have effect only on scroll position.

**Comment by Rich-Harris** [comment]
One way to think about it: if people are referencing `url.hash` in their `load` functions today, their apps are already broken (since `load` doesn't re-run when the hash changes). So it'd actually be fine for us to remove `url.hash` from `load`.

If we later decide that we _do_ need to make `url.hash` accessible in `load`, I would consider that a non-breaking change. So the semver-galaxy-brain thing to do is to make `url.hash` inaccessible in `load` (and fix the sticky `page` store issue in #4768), which coincidentally is by far the simplest thing to do, implementation-wise.

**Comment by jerrygreen** [comment]
Somehow even the simplest load function leads to `Error: url.hash is inaccessible from load. Consider accessing hash from the page store within the script tag of your component.`:

```ts
export const load: Load = (e) => {
  console.log(e)
}
```

Withouts this, I can't even see what actual data I'm receiving.

I mean, ofc there are docs etc, but cmon, svelte-kit is still experimental, things change and break often, googled answers lead to non-existing page docs, etc. I feel like it's easier for me to just print the actual data and see what happens, but somehow this error just completely stops my entire app. Couldn't it be a simple warning or something?
